refactor: refactor module

NotSugden opened the pull request:
First of all, I realise there are some major code style changes here, if the maintainers want i can revert some of those style changes

This PR does a major refactor to this module, here is a list of changes

**anyone who can help test this please do!, I have encountered issues with the bot jumping and placing a block underneath it, im not sure if this is an existing issue or something as a result of a change i've made**

- bumps the package version to `2.0.0` as there are major (breaking) changes in this PR
- adds sprinting funcionality
- adds (correct) typings
- refactors the injection function to:
  - renamed to `inject`
  - uses a class instead of how it was previously done
- code style changes:
  - ES6 arrow function syntax
  - uses JSDoc for comments on specific classes where applicable
  - refactors some functions to use built-in array methods like `Array.reduce` and `Array.some`

rom1504 commented:
> bumps the package version to 2.0.0 as there are major (breaking) changes in this PR

is this required ?

NotSugden commented:
> > bumps the package version to 2.0.0 as there are major (breaking) changes in this PR
> 
> is this required ?

well i suppose i could change the behaviour so its not breaking, but ultimately it feels better to just bump the major version instead, its up to you though

- alter this PR so no existing behaviour is broken
- keep the breaking changes and bump the major version

rom1504 commented:
What are the breaking changes and what's the benefit of them being breaking?
Introducing a major version has a big cost as it would require updates in all dependent packages and repos. So it makes sense to do that only if the breaking changes are worth it.

NotSugden commented:
> What are the breaking changes and what's the benefit of them being breaking?
> Introducing a major version has a big cost as it would require updates in all dependent packages and repos. So it makes sense to do that only if the breaking changes are worth it.

well i wasn't sure about every single breaking change as i didnt note them, but i'll compare everything and see

rom1504 commented:
I'd advise separating the new features and the refactoring changing nothing in 2 separate PRs, this looks hard to review

NotSugden commented:
@rom1504 

- bot.pathfinder is now a class, and not an object with various functions attached
- the inject function is now exported as `inject` and not `pathfinder`
- `bot` is now read-only and non-enumerable on the `Movements` class
- removed `pathfinder.isMining()` in favour of `pathfinder.mining`
- removed `pathfinder.isBuilding()` in favour of `pathfinder.building`
- removed `pathfinder.isThinking()` in favour of `pathfinder.thinking`
- removed the `done` callback parameter from `pathfinder.getPathTo()` as it was unnesercary
- `Movements#getBlock()` now returns its own object rather than directly modifying the `Block` recieved from `minecraft-data`
  - this change has lead to various parameter changes where a `Block` was previously accepted

NotSugden commented:
> I'd advise separating the new features and the refactoring changing nothing in 2 separate PRs, this looks hard to review

i'll look into doing this, although i don't think there are many intended new features in here, off the top of my head i added sprinting support (although that may warrant its own PR, i don't know if there are any side-effects from sprinting that will need to be addressed seperately)

Karang commented:
> @rom1504
> 
>     * bot.pathfinder is now a class, and not an object with various functions attached
> 
>     * the inject function is now exported as `inject` and not `pathfinder`
> 
>     * `bot` is now read-only and non-enumerable on the `Movements` class
> 
>     * removed `pathfinder.isMining()` in favour of `pathfinder.mining`
> 
>     * removed `pathfinder.isBuilding()` in favour of `pathfinder.building`
> 
>     * removed `pathfinder.isThinking()` in favour of `pathfinder.thinking`
> 
>     * removed the `done` callback parameter from `pathfinder.getPathTo()` as it was unnesercary
> 
>     * `Movements#getBlock()` now returns its own object rather than directly modifying the `Block` recieved from `minecraft-data`
>       
>       * this change has lead to various parameter changes where a `Block` was previously accepted

I don't get why you need to break the api?
I'm going to comment directly on the list of feature :
* what's the advantage of having pathfinder a class instead of a singleton object ?
* it was more explicit to be able to do `bot.loadPlugin(pathfinder)` instead of `bot.loadPlugin(inject)`, please revert
* what's the advantage of this ? Movement is an internal class anyway.
* this is an api break, what is the justification for it ?
* same
* same
* the done callback was intended to be used when implementing the async computation (if it takes more than a tick to compute)
* this was done like this for performance reasons: creating too many objects is expensive, recycling is better for this sensitive function

Keep it mind that this code is very performance critical, so when introducting that much changes, it will have to be benchmarked again and compared with the current version.

NotSugden commented:
@Karang 

- a class is alot easier to read and document than just a single object
- fair point, will revert
- i thought the `Movements` class is apart of the public API? the user instantiates it and the user changes properties on it to configure it ?
- well, seeing as the major version was going to be bumped due to other changes, i opted to remove this in favour of a getter as the properties exist on the class itself now there isn't a need for them to be a function, there isn't really a need for them to be aliased at all really
  - `pathfinder.isMining()` would be the same as just `pathfinder.digging`, however `pathfinder.isMining()` has been removed and now `pathfinder.mining` is just an alias for `pathfinder.digging`
- same
- same
- its not being used at the moment though ? shouldn't that have all gone into one PR? if im understanding what you're saying this was a WIP?
- but properties are being added to the `Block` object that shouldn't be there, couldn't this break behaviour if someone did say `Object.keys(minecraftData.blocksById[...])` (this is just an example), either way it seems a bit silly to me to do it that way, but if you do insist i can revert that

TheDudeFromCI commented on lib/goto.js:
```diff
@@ -13,23 +13,21 @@ function error (name, message) {
-  function cleanup (err) {
-    bot.removeListener('goal_reached', goalReached)
+  const cleanup = err => {
+    bot.removeListener('goal_reached', cleanup)
```
The goal reached event does pass a param to the callback, so using cleanup as the listener will wrongly throw an "error."

NotSugden replied:
oh, that is indeed correct, my bad

Karang commented:
> > I'd advise separating the new features and the refactoring changing nothing in 2 separate PRs, this looks hard to review
> 
> i'll look into doing this

You really should do what Rom suggested. I don't get why you are doing the "change style" part of this PR: as far as I'm concern there are some interesting stuff in this PR, but most of it is just shuffling things around for seemingly no reasons...

NotSugden commented:
> > > I'd advise separating the new features and the refactoring changing nothing in 2 separate PRs, this looks hard to review
> > 
> > 
> > i'll look into doing this
> 
> You really should do what Rom suggested. I don't get why you are doing the "change style" part of this PR: as far as I'm concerned there are some interesting stuff in this PR, but most of it is just shuffling things around for seemingly no reasons...
> 
> There is also a few changes that I think would slow down the pathfinding, this module was built with speed in mind, if you think your changes can make things faster, you'll have to run some benchmarks.

ok, ill make some branches on my fork and split up this PR a little bit, and then ill run some benchmarks to compare with the current version

Karang commented on examples/bench.js:
```diff
@@ -21,15 +20,29 @@ bot.loadPlugin(pathfinder)
-  console.log('Spawning took ' + (performance.now() - createTime).toFixed(2) + ' ms.')
+  console.log(
+    `Spawning took ${(performance.now() - createTime).toFixed(2)} ms.`
+  )
```
No need to make it spawn multiple lines.

Karang commented on examples/bench.js:
```diff
@@ -21,15 +20,29 @@ bot.loadPlugin(pathfinder)
+      process.exit()
+    },
+    10000
+  )
```
No need to make it spawn multiple lines.

Karang commented on examples/multiple.js:
```diff
@@ -1,8 +1,7 @@
+const { goals, pathfinder, Movements } = require('mineflayer-pathfinder')
 
-mineflayer.multiple = (bots, constructor) => {
+const multiple = (bots, constructor) => {
```
It was intended like this in prevision of moving this function to mineflayer. Please keep the namespace.

Karang commented on examples/multiple.js:
```diff
@@ -19,37 +18,53 @@ for (let i = 0; i < 40; i++) {
+          path.length
+        } moves. Computation took ${time.toFixed(2)} ms.`
+      )
+    )
```
No need to make it spawn multiple lines.

Karang commented on examples/multiple.js:
```diff
@@ -19,37 +18,53 @@ for (let i = 0; i < 40; i++) {
+        bot.pathfinder.setGoal(
+          new goals.GoalInvert(new goals.GoalFollow(target, 5)),
+          true
+        )
```
No need to make it spawn multiple lines.

Karang commented on examples/test.js:
```diff
@@ -23,59 +22,67 @@ bot.once('spawn', () => {
+      `I can get there in ${path.length} moves. Computation took ${time.toFixed(
+        2
+      )} ms (${nodesPerTick} nodes/tick).`
+    )
```
No need to make it spawn multiple lines.

Karang commented on lib/astar.js:
```diff
@@ -32,20 +28,23 @@ function reconstructPath (node) {
+      start,
+      0,
+      goal.heuristic(start)
+    ))
```
No need to make it spawn multiple lines.

Karang commented on lib/astar.js:
```diff
@@ -96,7 +95,12 @@ class AStar {
+          gFromThisNode,
+          this.goal.heuristic(neighborData),
+          node
+        )
```
No need to make it spawn multiple lines.

Karang commented on lib/astar.js:
```diff
@@ -110,4 +114,7 @@ class AStar {
+module.exports = {
+  AStar,
+  PathNode
+}
```
No need to make it spawn multiple lines.

Karang commented on lib/goals.js:
```diff
@@ -1,23 +1,36 @@
+  /**
+   * Return the distance between node and the goal
+   * @abstract
+   */
```
I prefer // over /**/ for comments as it makes it easier to comment codes when developing.
I'm also not a big fan of bulky comments.

Karang commented on lib/goals.js:
```diff
@@ -1,23 +1,36 @@
   hasChanged () {
     return false
-  }
+  };
```
Extra ;

Karang commented on lib/goals.js:
```diff
@@ -1,23 +1,36 @@
   isEnd (node) {
     return true
-  }
+  };
```
Extra;

Karang commented on lib/heap.js:
```diff
@@ -1,7 +1,5 @@
-
 class BinaryHeapOpenSet {
   constructor () {
-    // Initialing the array heap and adding a dummy element at index 0
```
Don't remove this comment.

Karang commented on lib/heap.js:
```diff
@@ -23,7 +21,10 @@ class BinaryHeapOpenSet {
     // Traversing up the parent node until the current node is greater than the parent
     while (current > 1 && this.heap[parent].f > this.heap[current].f) {
-      [this.heap[parent], this.heap[current]] = [this.heap[current], this.heap[parent]]
+      ;[this.heap[parent], this.heap[current]] = [
```
Extra ; No need to make it spawn multiple lines.

Karang commented on lib/move.js:
```diff
@@ -1,16 +1,27 @@
+    toBreak = [],
+    toPlace = [],
+    parkour = false
+  ) {
```
No need to make it spawn multiple lines.

Karang commented on lib/move.js:
```diff
@@ -1,16 +1,27 @@
 
-    this.hash = this.x + ',' + this.y + ',' + this.z
+  get hash () {
+    return `${this.x},${this.y},${this.z}`
```
The hash was precomputed for a reason.

Karang commented on lib/movements.js:
```diff
@@ -1,94 +1,92 @@
 class Movements {
-  constructor (bot, mcData) {
-    this.bot = bot
+  constructor (bot, minecraftData = _minecraftData(bot.version)) {
```
It is better to encourage people to give mcData rather than creating a new one behind their back. There are plans to have a single instance of mcData available in the bot, so this could be simplified, but not this way.

Karang commented on lib/movements.js:
```diff
@@ -1,94 +1,92 @@
     }
-    return b
+
+    if ('shapes' in block) {
```
This test is not necessary.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
       if (this.getBlock(block.position, 0, 0, 1).liquid) return false
     }
-    return block.type && !this.blocksCantBreak.has(block.type)
-    // TODO: break exclusion areas
```
Don't remove this comment

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+      false,
+      enchants,
+      effects
+    )
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+          dx: dir.x,
+          dy: 0,
+          dz: dir.z
+        })
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+        dx: 0,
+        dy: 1,
+        dz: 0
+      })
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+        toBreak,
+        toPlace
+      )
+    )
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+        dx: dir.x,
+        dy: 0,
+        dz: dir.z
+      })
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+        toBreak,
+        toPlace
+      )
+    )
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+        cost,
+        toBreak
+      )
+    )
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+        cost,
+        toBreak,
+        toPlace
+      )
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+        toBreak,
+        toPlace
+      )
+    )
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+      dy: 1,
+      dz: 0,
+      jump: true
+    })
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+        toBreak,
+        toPlace
+      )
+    )
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+          true
+        )
+      )
+    } else if (safeBlockB.safe && safeBlockC.physical) {
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+          true
+        )
+      )
+    } else if (safeBlockC.safe && safeBlockD.safe) {
```
No need to make it spawn multiple lines.

Karang commented on lib/movements.js:
```diff
@@ -104,213 +102,365 @@ class Movements {
+          [],
+          true
+        )
+      )
```
No need to make it spawn multiple lines.

Karang commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "mineflayer-pathfinder",
-  "version": "1.0.12",
+  "version": "2.0.0",
```
Not a major version + version should only be increased in dedicated commit.

Karang commented on readme.md:
```diff
@@ -49,27 +47,29 @@ bot.once('spawn', () => {
+- Can break/place blocks as part of its deplacement
+- Automatically update path when environment change
+- Long distance paths
+- Can swim
```
Not needed.

Karang commented on examples/bench.js:
```diff
@@ -21,15 +20,29 @@ bot.loadPlugin(pathfinder)
-    bot.quit()
-    process.exit()
-  }, 10000)
+  const goal = new goals.GoalXZ(
```
I prefer to import explicitly which goals get used and use `new GoalXZ`

Karang reviewed (changes requested):
Can you apply the following changes? It will make the diff clearer and easier to review.

Hopefully, we don't have an 80 character limit in modern IDE, so there is plenty of horizontal space of the screen we can use to avoid scrolling up/down for ages ;)
